dqn.py

```diff
@@ -0,0 +1,259 @@
+        
+
+    def fill_buffer(self, min_buffer_size):
+        s = self.env.reset()
+        for _ in range(min_buffer_size):
+            a = random.randrange(self.env.num_actions)
+
+            s1, r, done = self.env.step(a)
+            self.buffer.add(s, a, s1, r, done)
+
+            if done:
+                s = self.env.reset()
+
+
+    def get_action(self, s, step):
+        # Get random action if under epsilon otherwise take from action value
+        e = np.interp(step, [0, self.e_decay], [self.e_start, self.e_end])
+        sample = random.random()
+        if sample <= e:
+            a = random.randrange(self.env.num_actions)
+            return a
+        else:
+            a = self.Q.act(s)
+            return a 
+
+
+    def get_return(self, reward_list):
+        Return = reward_list.pop(0)
+        for i in range(len(reward_list)):
+            Return += (GAMMA**i) * reward_list[i]
+        return Return
+                
+
+    def optimize(self, step):
+        transitions = self.buffer.sample(BATCH_SIZE)
+        batch = Transition(*zip(*transitions))
+
+        # make training batches        
+        state_batch = np.array(batch.state)
+        action_batch = np.asarray(batch.action)
+        next_state_batch = np.asarray(batch.next_state)
+        reward_batch = np.asarray(batch.reward)
+        dones_batch = np.asarray(batch.done)
+
+        state_batch_t = torch.as_tensor(state_batch, dtype=torch.float32)
+        action_batch_t = torch.as_tensor(action_batch, dtype=torch.int64).unsqueeze(-1)
+        next_state_batch_t = torch.as_tensor(next_state_batch, dtype=torch.float32)
+        reward_batch_t = torch.as_tensor(reward_batch, dtype=torch.float32).unsqueeze(-1)
+        dones_batch_t = torch.as_tensor(dones_batch, dtype=torch.float32).unsqueeze(-1)
+        
+        
+        # compute targets
+        target_q_vals = self.target_Q(next_state_batch_t)
+        max_target_q_vals = target_q_vals.max(dim=1, keepdim=True)[0]
+        targets = reward_batch_t + GAMMA * (1 - dones_batch_t) * max_target_q_vals
+
+        # get action q vals
+        q_vals = self.Q(state_batch_t)
+        action_q_vals = torch.gather(input=q_vals, dim=1, index=action_batch_t)
+
+        criterion = nn.SmoothL1Loss()
+        loss = criterion(action_q_vals, targets)
+        self.optimizer.zero_grad()
+        loss.backward()
+        self.optimizer.step()
+
+        # Update target net
+        if step % 100 == 0:
+            self.target_Q.load_state_dict(self.Q.state_dict())
+
+        if step % 100 == 0:
+            print()
+            print('Step:', step)
+            print('Avg Reward:', np.mean(self.rewards))
+
+        
+
+    def train(self):
+        self.fill_buffer(MIN_BUFFER_SIZE)
+
+        s = self.env.reset()
+        epi_reward = 0.0
+        for step in itertools.count():
+            a = self.get_action(s, step)
+
+            s1, r, done = self.env.step(a)
+            self.buffer.add( s, a, s1, r, done)
```

Comment:
Why do you have to fill your buffer with random values before adding the sampled values?